doc/langref.html.in

```diff
@@ -5088,6 +5088,46 @@ fn deferUnwindExample() void {
 
 test "defer unwinding" {
     deferUnwindExample();
+}
+      {#code_end#}
+      {#code_begin|test|test_mutable_vars_defer#}
+const std = @import("std");
+
+const expect = std.testing.expect;
+
+// It's important to understand the semantics of `defer` when using mutable
+// variables. The `defer` statement captures the value of the variable at the
+// time the `defer` statement is executed, which is at the end of the current
+// scope. This means that if the variable is mutated after the `defer` statement
+// is defined but before the end of the scope, the `defer` statement will
+// execute with the mutated value.
+test "defer with mutable variables" {
+    var x: u8 = 42;
+    defer expect(x == 69) catch unreachable;
+    x += 27;
+}
+
+// Caller owns the slice.
+fn getSomeText(allocator: std.mem.Allocator) ![]u8 {
+    var buffer = std.ArrayList(u8).init(allocator);
+    try buffer.appendSlice(" Hello, world!\n");
+    return try buffer.toOwnedSlice();
+}
+
+test "defer with mutable variables implications" {
```

Comment:
In my opinion, this example is too contrived because it relies on another function and requires the reader to know what `std.mem.trim` does. I would prefer a simpler example of correct code, like [the one mlugg wrote](<https://github.com/ziglang/zig/issues/18490#issuecomment-1882089070>), or none at all, since this feature of `defer` is already covered above.

Reply:
@notcancername I disagree with this. The main reason I'm updating the documentation is to help document a potential footgun, so I think it's good to have an example that explicitly showcases the exact type of problem you can run into. Focusing on the happy path doesn't help here. Point taken about the `std.mem.trim()` usage though, so maybe I can make that part clearer?

Reply:
You could slice `text` explicitly: 
```
text = text[1..];
```

Would you be alright with changing the external function usage to `std.testing.allocator.dupeZ("Hello, world!")`?

Reply:
@notcancername Yes, good ideas. You happen to have a more complete example? (If not, it's fine. I'm on mobile right now, so I'll make the updates once I'm back at my computer.)